File: src/test/java/com/dndweapons/catalog/WeaponSpecTest.java
package com.dndweapons.catalog;

import org.junit.jupiter.api.Test;

import java.util.Set;

import static org.junit.jupiter.api.Assertions.*;

class WeaponSpecTest {

    @Test
    void canConstructMinimalMeleeSpec() {
        WeaponSpec longsword = new WeaponSpec(
            "longsword", "Longsword",
            Category.MARTIAL_MELEE, DamageType.SLASHING,
            "1d8", "1d10",
            6, 1, 1.5f, 0.0f, 0,
            Set.of(Property.VERSATILE),
            RangeKind.NONE, 250, null
        );
        assertEquals("longsword", longsword.id());
        assertEquals(6, longsword.attackDamage());
        assertTrue(longsword.properties().contains(Property.VERSATILE));
    }

    @Test
    void blankIdIsRejected() {
        assertThrows(IllegalArgumentException.class, () -> new WeaponSpec(
            "", "Longsword",
            Category.MARTIAL_MELEE, DamageType.SLASHING,
            "1d8", null,
            6, 0, 1.5f, 0.0f, 0,
            Set.of(),
            RangeKind.NONE, 250, null
        ));
    }

    @Test
    void vanillaRoleTagWhenSetMakesItVanillaMapped() {
        WeaponSpec shortsword = new WeaponSpec(
            "shortsword", "Shortsword",
            Category.MARTIAL_MELEE, DamageType.PIERCING,
            "1d6", null,
            5, 0, 1.8f, 0.0f, 0,
            Set.of(Property.LIGHT, Property.FINESSE),
            RangeKind.NONE, 250,
            "dndweapons:role/shortsword"
        );
        assertNotNull(shortsword.vanillaRoleTag());
        assertTrue(shortsword.isVanillaMapped());
    }

    @Test
    void nullVanillaRoleTagMeansRegisteredItem() {
        WeaponSpec longsword = new WeaponSpec(
            "longsword", "Longsword",
            Category.MARTIAL_MELEE, DamageType.SLASHING,
            "1d8", "1d10",
            6, 1, 1.5f, 0.0f, 0,
            Set.of(Property.VERSATILE),
            RangeKind.NONE, 250, null
        );
        assertNull(longsword.vanillaRoleTag());
        assertFalse(longsword.isVanillaMapped());
    }
}

File: src/test/kotlin/com/dndweapons/catalog/WeaponSpecTest.kt
package com.dndweapons.catalog

import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertFalse
import org.junit.jupiter.api.Assertions.assertNotNull
import org.junit.jupiter.api.Assertions.assertNull
import org.junit.jupiter.api.Assertions.assertThrows
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test

class WeaponSpecTest {

    @Test
    fun canConstructMinimalMeleeSpec() {
        val longsword = WeaponSpec(
            id = "longsword", displayName = "Longsword",
            category = Category.MARTIAL_MELEE, damageType = DamageType.SLASHING,
            diceText = "1d8", versatileDice = "1d10",
            attackDamage = 6, versatileBonus = 1,
            attackSpeed = 1.5f, reachBonus = 0.0f, knockbackBonus = 0,
            properties = setOf(Property.VERSATILE),
            ranged = RangeKind.NONE, baseDurability = 250,
            vanillaRoleTag = null,
        )
        assertEquals("longsword", longsword.id)
        assertEquals(6, longsword.attackDamage)
        assertTrue(longsword.properties.contains(Property.VERSATILE))
    }

    @Test
    fun blankIdIsRejected() {
        assertThrows(IllegalArgumentException::class.java) {
            WeaponSpec(
                id = "", displayName = "Longsword",
                category = Category.MARTIAL_MELEE, damageType = DamageType.SLASHING,
                diceText = "1d8", versatileDice = null,
                attackDamage = 6, versatileBonus = 0,
                attackSpeed = 1.5f, reachBonus = 0.0f, knockbackBonus = 0,
                properties = emptySet(),
                ranged = RangeKind.NONE, baseDurability = 250,
                vanillaRoleTag = null,
            )
        }
    }

    @Test
    fun vanillaRoleTagWhenSetMakesItVanillaMapped() {
        val shortsword = WeaponSpec(
            id = "shortsword", displayName = "Shortsword",
            category = Category.MARTIAL_MELEE, damageType = DamageType.PIERCING,
            diceText = "1d6", versatileDice = null,
            attackDamage = 5, versatileBonus = 0,
            attackSpeed = 1.8f, reachBonus = 0.0f, knockbackBonus = 0,
            properties = setOf(Property.LIGHT, Property.FINESSE),
            ranged = RangeKind.NONE, baseDurability = 250,
            vanillaRoleTag = "dndweapons:role/shortsword",
        )
        assertNotNull(shortsword.vanillaRoleTag)
        assertTrue(shortsword.isVanillaMapped)
    }

    @Test
    fun nullVanillaRoleTagMeansRegisteredItem() {
        val longsword = WeaponSpec(
            id = "longsword", displayName = "Longsword",
            category = Category.MARTIAL_MELEE, damageType = DamageType.SLASHING,
            diceText = "1d8", versatileDice = "1d10",
            attackDamage = 6, versatileBonus = 1,
            attackSpeed = 1.5f, reachBonus = 0.0f, knockbackBonus = 0,
            properties = setOf(Property.VERSATILE),
            ranged = RangeKind.NONE, baseDurability = 250,
            vanillaRoleTag = null,
        )
        assertNull(longsword.vanillaRoleTag)
        assertFalse(longsword.isVanillaMapped)
    }
}
